test/integration/scheduler_perf/scheduler_bench_test.go

```diff
@@ -17,44 +17,87 @@ limitations under the License.
 	"k8s.io/kubernetes/test/integration/framework"
 	testutils "k8s.io/kubernetes/test/utils"
 
 	"github.com/golang/glog"
 )
 
-// BenchmarkScheduling100Nodes0Pods benchmarks the scheduling rate
-// when the cluster has 100 nodes and 0 scheduled pods
-func BenchmarkScheduling100Nodes0Pods(b *testing.B) {
-	benchmarkScheduling(100, 0, b)
+// BenchmarkScheduling benchmarks the scheduling rate when the cluster has
+// various quantities of nodes and scheduled pods.
+func BenchmarkScheduling(b *testing.B) {
+	tests := []struct{ nodes, pods int }{
+		{nodes: 100, pods: 0},
+		{nodes: 100, pods: 1000},
+		{nodes: 1000, pods: 0},
+		{nodes: 1000, pods: 1000},
+	}
+	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc1")
+	testStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc2")
+	for _, test := range tests {
+		name := fmt.Sprintf("%vNodes/%vPods", test.nodes, test.pods)
+		b.Run(name, func(b *testing.B) { benchmarkScheduling(test.nodes, test.pods, setupStrategy, testStrategy, b) })
+	}
 }
 
-// BenchmarkScheduling100Nodes1000Pods benchmarks the scheduling rate
-// when the cluster has 100 nodes and 1000 scheduled pods
-func BenchmarkScheduling100Nodes1000Pods(b *testing.B) {
-	benchmarkScheduling(100, 1000, b)
-}
+// BenchmarkSchedulingAntiAffinity benchmarks the scheduling rate when the
+// cluster has various quantities of nodes and scheduled pods.
+// New pods are scheduled with anti-affinity matching the pre-existing pods.
+func BenchmarkSchedulingAntiAffinity(b *testing.B) {
+	tests := []struct{ nodes, pods int }{
+		{nodes: 500, pods: 50},
+		{nodes: 500, pods: 400},
+	}
+	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("rc")
+	// NewSimpleWithControllerCreatePodStrategy will add name=rc to pods.
+	basePod := makeBasePodWithAntiAffinity(map[string]string{"name": "rc"})
+	testStrategy := testutils.NewCustomCreatePodStrategy(basePod)
+	for _, test := range tests {
+		name := fmt.Sprintf("%vNodes/%vPods", test.nodes, test.pods)
+		b.Run(name, func(b *testing.B) { benchmarkScheduling(test.nodes, test.pods, setupStrategy, testStrategy, b) })
+	}
 
-// BenchmarkScheduling1000Nodes0Pods benchmarks the scheduling rate
-// when the cluster has 1000 nodes and 0 scheduled pods
-func BenchmarkScheduling1000Nodes0Pods(b *testing.B) {
-	benchmarkScheduling(1000, 0, b)
 }
 
-// BenchmarkScheduling1000Nodes1000Pods benchmarks the scheduling rate
-// when the cluster has 1000 nodes and 1000 scheduled pods
-func BenchmarkScheduling1000Nodes1000Pods(b *testing.B) {
-	benchmarkScheduling(1000, 1000, b)
+// makeBasePodWithAntiAffinity creates a Pod object to be used as a template.
+// The Pod has an anti-affinity requirement against nodes running pods with the given labels.
+func makeBasePodWithAntiAffinity(labels map[string]string) *v1.Pod {
+	basePod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "affinity-pod-",
+			Labels:       map[string]string{"testType": "affinity"},
```

Comment:
Would it be better to have matching labels in our (anti)-affinity rule?

Reply:
I wasn't sure here. My thought was that all the pods scheduled during benchmark have anti-affinity to the pre-scheduled pods, but not each other. Do you think the pods should also have anti-affinity with themselves?

The testType label isn't serving any purpose other than possibly debugging. Maybe just omit it?

Reply:
I don't see any config that specifies anti-affinity to existing pods. Given that the number of pods in your tests is smaller than or equal to the number of nodes, I think it would make the tests better if we added anti-affinity among the created pods.

Reply:
Per our discussion, I've added a lower bound to b.N (and rebased on top of that commit). I also updated the tests' PodAntiAffinity rules in the second fixup (27d1ce3).

Long term, I think we will end up moving away from micro-benchmarks for this sort of test.